File: samples/afas-openfhir-dotnet/afas-sample-usecases/OrganizationTest.cs
using NUnit.Framework;
using Hl7.Fhir.Model;
using System;
using Hl7.Fhir.Rest;


namespace sample_usecases_tests
{
    class OrganizationTest
    {

        /**
         * Search all organizations
         */
        [Test]
        public void SearchAll()
        {
            // create the client:
            var client = FhirTestUtils.CreateClient();

            var bundle = client.Search<Organization>();
            foreach (var be in bundle.Entry)
            {
                // print ids:
                var organization = be.Resource as Organization;
                Console.WriteLine($"Organization found: id={organization.IdElement.Value} name={organization.Name}");
            }
        }



        /**
         * Search by lastÛpdated
         */
        [Test]
        public void SearchModifiedFrom()
        {
            // create the client:
            var client = FhirTestUtils.CreateClient();

            var q = new SearchParams()
              .Where("_lastUpdated=ge2022-08-05T14:51:04")
              .LimitTo(50);
            var bundle = client.Search<Organization>(q);
            foreach (var be in bundle.Entry)
            {
                // print ids:
                var organization = be.Resource as Organization;
                Console.WriteLine($"Organization found: id={organization.IdElement.Value} lastUpdate={organization.Meta.LastUpdated}");
            }
        }

        /**
         * Search by identifiers
         */
        [Test]
        public void SearchByIdentifiers()
        {
            // create the client:
            var client = FhirTestUtils.CreateClient();

            var q = new SearchParams()
              .Where("identifier=001604103000,01603998400,001604252500")
              .LimitTo(50);
            var bundle = client.Search<Organization>(q);
            foreach (var be in bundle.Entry)
            {
                // print ids:
                var organization = be.Resource as Organization;
                Console.WriteLine($"Organization found: id={organization.IdElement.Value}");
            }
        }

        /**
         * Search by Finess number
         */
        [Test]
        public void SearchByFinessNumber()
        {
            // create the client:
            var client = FhirTestUtils.CreateClient();

            var q = new SearchParams()
              .Where("identifier=https://finess.esante.gouv.fr|010000602,https://finess.esante.gouv.fr|010000628")
              .LimitTo(50);
            var bundle = client.Search<Organization>(q);
            foreach (var be in bundle.Entry)
            {
                // print ids:
                var organization = be.Resource as Organization;
                Console.WriteLine($"Organization found: id={organization.IdElement.Value}");
            }
        }
        
        /**
         * Search by type
         */
        [Test]
        public void SearchByType()
        {
            // create the client:
            var client = FhirTestUtils.CreateClient();

            var q = new SearchParams()
              .Where("type=http://interopsante.org/fhir/CodeSystem/fr-v2-3307|GEOGRAPHICAL-ENTITY")
              .LimitTo(50);
            var bundle = client.Search<Organization>(q);
            foreach (var be in bundle.Entry)
            {
                // print ids:
                var organization = be.Resource as Organization;
                var types = "";
                foreach(var type in organization.Type)
                {
                    types = types + " - " +type.Coding[0].Code;
                }
                Console.WriteLine($"Organization found: id={organization.IdElement.Value} type={types}");
            }
        }


        /**
         * Search by type (INSEE)
         */
        [Test]
        public void SearchByINSEENomenclature()
        {
            // create the client:
            var client = FhirTestUtils.CreateClient();

            var q = new SearchParams()
              .Where("type=https://mos.esante.gouv.fr/NOS/TRE_R75-InseeNAFrev2Niveau5/FHIR/TRE-R75-InseeNAFrev2Niveau5|82.19Z")
              .LimitTo(50);
            var bundle = client.Search<Organization>(q);
            foreach (var be in bundle.Entry)
            {
                // print ids:
                var organization = be.Resource as Organization;
                var types = "";
                foreach (var type in organization.Type)
                {
                    types = types + " - " + type.Coding[0].Code;
                }
                Console.WriteLine($"Organization found: id={organization.IdElement.Value} type={types}");
            }
        }



        /**
         * Search by activity type
         */
        [Test]
        public void SearchByActivity()
        {
            // create the client:
            var client = FhirTestUtils.CreateClient();

            var q = new SearchParams()
              .Where("type=https://mos.esante.gouv.fr/NOS/TRE_R02-SecteurActivite/FHIR/TRE-R02-SecteurActivite|SA29")
              .LimitTo(50);
            var bundle = client.Search<Organization>(q);
            foreach (var be in bundle.Entry)
            {
                // print ids:
                var organization = be.Resource as Organization;
                var types = "";
                foreach (var type in organization.Type)
                {
                    types = types + " - " + type.Coding[0].Code;
                }
                Console.WriteLine($"Organization found: id={organization.IdElement.Value} activity={types}");
            }
        }
        
        /**
         * Search by name
         */
        [Test]
        public void SearchByNameTwoTerms()
        {
            // create the client:
            var client = FhirTestUtils.CreateClient();

            var q = new SearchParams()
              .Where("name:contains=imagerie").Add("name:contains", "centre")
              .LimitTo(50);
            var bundle = client.Search<Organization>(q);
            foreach (var be in bundle.Entry)
            {
                // print ids:
                var organization = be.Resource as Organization;
                Console.WriteLine($"Organization found: name={organization.Name}");
            }
        }
                
        /**
         * Search by zip code
         */
        [Test]
        public void SearchByZipCode()
        {
            // create the client:
            var client = FhirTestUtils.CreateClient();

            var q = new SearchParams()
              .Where("address-postalcode=13290,13290")
              .LimitTo(50);
            var bundle = client.Search<Organization>(q);
            foreach (var be in bundle.Entry)
            {
                // print ids:
                var organization = be.Resource as Organization;
                Console.WriteLine($"Organization found: name={organization.Name} | zipCode={organization.Address[0].PostalCode}");
            }
        }

    }
}

File: samples/asaf-openfhir-java/modules/api-usage-tests/src/test/java/fr/ans/asaf/demo/OrganizationTest.java
package fr.ans.asaf.demo;

import ca.uhn.fhir.rest.gclient.DateClientParam;
import ca.uhn.fhir.rest.gclient.StringClientParam;
import org.hl7.fhir.r4.model.Bundle;
import org.hl7.fhir.r4.model.Organization;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.stream.Collectors;

/**
 * Samples in this file show how to search Organizations
 *
 * @author Guillaume Poulériguen
 * @since 1.0.0
 */
public class OrganizationTest {

    private static final Logger logger = LoggerFactory.getLogger(OrganizationTest.class);

    /**
     * Search all organizations
     */
    @Test
    public void searchAll() {
        // create the client:
        var client = FhirTestUtils.createClient();
        var bundle = client.search().forResource(Organization.class).returnBundle(Bundle.class).execute();
        for (var organizationEntry : bundle.getEntry()) {
            // print Organization ids:
            var organization = (Organization) organizationEntry.getResource();
            logger.info("Organization found: id={} name={}", organization.getIdElement().getIdPart(), organization.getName());
        }
    }

    @Test
    public void searchModifiedFrom() {
        // create the client:
        var client = FhirTestUtils.createClient();

        // create the date search parameter :
        var dateParam = new DateClientParam("_lastUpdated");

        var bundle = client.search()
                .forResource(Organization.class)
                .where(dateParam.afterOrEquals().second("2022-08-05T14:51:04"))
                .returnBundle(Bundle.class).execute();

        for (var organizationEntry : bundle.getEntry()) {
            // cast entry :
            var organization = (Organization) organizationEntry.getResource();
            // print update date & id :
            logger.info("Organization found: id={} lastUpdate={}", organization.getIdElement().getIdPart(), organization.getMeta().getLastUpdated());
        }
    }

    @Test
    public void searchByIdentifiers() {
        // create the client:
        var client = FhirTestUtils.createClient();

        // create the identifier search parameter :
        var idParam = new StringClientParam("identifier");

        var bundle = client.search()
                .forResource(Organization.class)
                .where(idParam.matches().values("001604103000", "01603998400", "001604252500"))
                .returnBundle(Bundle.class).execute();

        for (var organizationEntry : bundle.getEntry()) {
            // cast entry :
            var organization = (Organization) organizationEntry.getResource();
            // print id :
            logger.info("Organization found: id={}", organization.getIdentifierFirstRep().getValue());
        }
    }

    @Test
    public void searchByFinessNumber() {
        // create the client:
        var client = FhirTestUtils.createClient();

        // create finess where clause
        var finessSearchClause = Organization.IDENTIFIER.exactly().systemAndValues(
                "https://finess.esante.gouv.fr", "010000602", "010000628", "010000735");

        var bundle = client.search()
                .forResource(Organization.class)
                .where(finessSearchClause)
                .returnBundle(Bundle.class).execute();

        for (var organizationEntry : bundle.getEntry()) {
            // cast entry :
            var organization = (Organization) organizationEntry.getResource();
            // print id :
            logger.info("Organization found: id={}", organization.getIdentifierFirstRep().getValue());
        }
    }

    @Test
    public void searchByType() {
        // create the client:
        var client = FhirTestUtils.createClient();

        // create the type search parameter :
        var typeSearchClause = Organization.TYPE.exactly().systemAndValues(
                "http://interopsante.org/fhir/CodeSystem/fr-v2-3307", "GEOGRAPHICAL-ENTITY");

        var bundle = client.search()
                .forResource(Organization.class)
                .where(typeSearchClause)
                .returnBundle(Bundle.class).execute();

        for (var organizationEntry : bundle.getEntry()) {
            // cast entry :
            var organization = (Organization) organizationEntry.getResource();
            // print id :
            var organizationCodes = organization.getType().stream().map(type -> type.getCodingFirstRep().getCode()).collect(Collectors.joining(" - "));
            logger.info("Organization found: name={} type={}", organization.getName(), organizationCodes);
        }
    }


    @Test
    public void searchByINSEENomenclature() {
        // create the client:
        var client = FhirTestUtils.createClient();

        // create the search parameter :
        var typeSearchClause = Organization.TYPE.exactly().systemAndValues(
                "https://mos.esante.gouv.fr/NOS/TRE_R75-InseeNAFrev2Niveau5/FHIR/TRE-R75-InseeNAFrev2Niveau5", "82.19Z");

        var bundle = client.search()
                .forResource(Organization.class)
                .where(typeSearchClause)
                .returnBundle(Bundle.class).execute();

        for (var organizationEntry : bundle.getEntry()) {
            // cast entry :
            var organization = (Organization) organizationEntry.getResource();
            // print data :
            var organizationCodes = organization.getType().stream().map(type -> type.getCodingFirstRep().getCode()).collect(Collectors.joining(" - "));
            logger.info("Organization found: id={} type={}", organization.getName(), organizationCodes);
        }
    }

    @Test
    public void searchByActivity() {
        // create the client:
        var client = FhirTestUtils.createClient();

        // create the activity search parameter :
        var activitySearchClause = Organization.TYPE.exactly().systemAndValues(
                "https://mos.esante.gouv.fr/NOS/TRE_R02-SecteurActivite/FHIR/TRE-R02-SecteurActivite", "SA29");

        var bundle = client.search()
                .forResource(Organization.class)
                .where(activitySearchClause)
                .returnBundle(Bundle.class).execute();

        for (var organizationEntry : bundle.getEntry()) {
            // cast entry :
            var organization = (Organization) organizationEntry.getResource();
            // print data :
            var activityName = organization.getTypeFirstRep().getCodingFirstRep().getDisplay();
            logger.info("Organization found: name={} activity={}", organization.getName(), activityName);
        }
    }

    @Test
    public void searchByNameTwoTerms() {
        // create the client:
        var client = FhirTestUtils.createClient();

        // create the name search parameter :
        var nameSearchClause = Organization.NAME.contains().values("imagerie", "centre");

        var bundle = client.search()
                .forResource(Organization.class)
                .where(nameSearchClause)
                .returnBundle(Bundle.class).execute();

        for (var organizationEntry : bundle.getEntry()) {
            // cast entry :
            var organization = (Organization) organizationEntry.getResource();
            // print data :
            logger.info("Organization found: name={}", organization.getName());
        }
    }

    @Test
    public void searchByZipCode() {
        // create the client:
        var client = FhirTestUtils.createClient();

        // create the postal code search parameter :
        var nameSearchClause = Organization.ADDRESS_POSTALCODE.matchesExactly().values("13290", "13321");

        var bundle = client.search()
                .forResource(Organization.class)
                .where(nameSearchClause)
                .returnBundle(Bundle.class).execute();

        for (var organizationEntry : bundle.getEntry()) {
            // cast entry :
            var organization = (Organization) organizationEntry.getResource();
            // print data :
            logger.info("Organization found: name={} | zipCode={}", organization.getName(), organization.getAddressFirstRep().getPostalCode());
        }
    }


}
